units/log_output_tests.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstddef>
+#include <cstdio>
+#include <fstream>
+#include <string>
+#include <thread>
+
+#include "debug_logging.h"
+#include "gtest/gtest.h"
+
+#define LOGS_PER_THREAD 1337
+#define THREAD_COUNT 100
+
+namespace performancelayers {
+namespace {
+// An abstraction over the output. It writes the incoming string to the output.
+class LogOutput {
+ public:
+  virtual ~LogOutput() = default;
+
+  virtual void LogLine(const std::string &line) = 0;
+};
+
+// Implements LogOutput for a file. If the given filename is `nullptr`, it
+// writes to the standard output. After each write, the logs are persisted to
+// the file by calling `fflush`. Since only one line is written to the output by
+// each `LogLine()` call, it doens't need to aquire a mutex.
+class FileOutput : public LogOutput {
+ public:
+  FileOutput(const char *filename) {
+    if (filename) {
```

Comment:
nit: prefer early exits, e.g.:
```c++
if (!filename) {
  out_ = stderr;
  return;
}

out_ = fopen(...);
```

See https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code. There's also a similar rule made specifically for the google c++ coding style guide but I can't find it at the moment (should be one of https://abseil.io/tips/)

Reply:
Thanks! It was really helpful!